pysoa/server/types.py

```diff
@@ -24,5 +44,68 @@ class EnrichedActionRequest(ActionRequest):
     context = attr.ib(default=attr.Factory(dict))
     control = attr.ib(default=attr.Factory(dict))
     client = attr.ib(default=None)
-    async_event_loop = attr.ib(default=None)
-    run_coroutine = attr.ib(default=None)
+    async_event_loop = attr.ib(default=None)  # deprecated
+    run_coroutine = attr.ib(default=None)  # replacement for async_event_loop
+
+    def call_local_action(self, action, body, raise_action_errors=True):
+        # type: (six.text_type, Dict, bool) -> ActionResponse
+        """
+        This helper calls another action, locally, that resides on the same service, using the provided action name
+        and body. The called action will receive a copy of this request object with different action and body details.
+
+        :param action: The action to call (must exist within the `action_class_map` from the `Server` class)
+        :type action: union[str, unicode]
+        :param body: The body to send to the action
+        :type body: dict
+        :param raise_action_errors: If `True` (the default), all action errors will be raised; otherwise, an
+                                    `ActionResponse` containing the errors will be returned.
+        :type raise_action_errors: bool
+
+        :return: the action response.
+        :rtype: ActionResponse
+
+        :raises: ActionError
+        """
+        server = getattr(self, '_server', None)
+        if not server:
+            errors = [Error(
+                code=ERROR_CODE_SERVER_ERROR,
+                message="No `_server` attribute set on action request object (and can't make request without it)",
+            )]
+            if raise_action_errors:
+                raise ActionError(errors)
+            return ActionResponse(action=action, errors=errors)
+
+        if action not in server.action_class_map:
+            errors = [Error(
+                code=ERROR_CODE_UNKNOWN,
+                message='The action "{}" was not found on this server.'.format(action),
+                field='action',
+            )]
+            if raise_action_errors:
+                raise ActionError(errors)
+            return ActionResponse(action=action, errors=errors)
+
+        action_callable = (
+            server.action_class_map[action](server.settings)
+        )  # type: Callable[[ActionRequest], ActionResponse]
+
+        request = self.__class__(
+            action=action,
+            body=body,
+            # Dynamically copy all Attrs attributes so that subclasses introducing other Attrs can still work properly
+            **{a.name: getattr(self, a.name) for a in self.__attrs_attrs__ if a.name not in ('action', 'body')}
+        )
+        request._server = server
+
+        try:
+            response = action_callable(request)
```

Comment:
Should middleware be run here?  What should users of this expect if they are used to, for example, a metrics middleware to be running on their actions normally?  It would be good to have clearer documentation on the differences (if any) in how this local action would run from an externally triggered action and what environmental expectations are appropriate.

Reply:
Great question, @adamsussman. I actually looked at doing this at first. There are some complications involved, and I would need to proceed carefully. I'm not sure I know the exact right way to do it yet. Critically, I looked through all of our internal middleware and I don't yet see any use cases where running the middleware here would make something work that otherwise wouldn't work, except maybe that we wouldn't record metrics around this action (which might even be desirable) (but the metrics container will still be passed in to the action, so it can still record _custom_ metrics). I'm not ruling out adding middleware in here in the future, but I'm not sure it's necessary _right now_. However, if you believe it is necessary, I can spend some more time on this now instead of putting that off.

Keep in mind that the way we are currently doing this (manually instantiating the class and calling the action) also does not run any middleware, so we aren't taking a step back from that. This is actually an improvement over our current process. Keep in mind, also, that there are other things we are skipping here: server metrics, request logging, etc. We're not treating this as a request. It's basically a helper to locally use the action class directly. For whatever that is worth.

Reply:
I'm on the fence on whether its necessary.  If nothing else though, clear documentation on exactly what is or is not happening would be good.

Reply:
Fair enough. I have expanded the documentation to specifically explain the things that do not happen and clarify that we may add middleware execution to the helper in the future.